mondrian/src/main/java/mondrian/rolap/RolapSchemaPool.java

```diff
@@ -371,10 +372,43 @@ private static String getSchemaContent(
                 }
             }
         }
+        else {
+            // If exist a DynamicSchemaProcessor apply locale
+            if ( !Util.isEmpty( dynProcName ) ) {
+                return processDynamicSchemaWithCatalog( dynProcName, catalogStr, connectInfo );
+            }
+        }
 
         return catalogStr;
     }
 
+    private static String processDynamicSchemaWithCatalog(
+      final String dynProcName,
+      final String catalogStr,
+      final Util.PropertyList connectInfo) {
+
+        String catalogStrProcessed = catalogStr;
+
+        if (RolapSchema.LOGGER.isDebugEnabled()) {
+            RolapSchema.LOGGER.debug(
+              "Pool.get: translate schema \"" + catalogStr
+                + "\" using dynamic processor");
+        }
+        try {
+            final DynamicSchemaProcessor dynProc =
+              ClassResolver.INSTANCE.instantiateSafe(dynProcName);
+            if( dynProc instanceof LocalizingDynamicSchemaProcessor )
+                catalogStrProcessed = ((LocalizingDynamicSchemaProcessor)dynProc).filter( catalogStrProcessed, connectInfo);
```

Comment:
Do you really need to cast to localizing processor class? I understand that you created a new method in the localizing class, but you should only do this as a last resort. It breaks the API. The processor's API should be sufficient. You can use the filter method that already exists. There's no need to check for instanceof either.

https://github.com/pentaho/mondrian/blob/master/mondrian/src/main/java/mondrian/spi/DynamicSchemaProcessor.java#L53

Reply:
The interface DynamicSchemaProcessor.java only have the method: 
`public String processSchema( String schemaUrl, Util.PropertyList connectInfo) throws Exception;` 

This method is implemented by FilterDynamicSchemaProcessor that it only creates an InputStream from schemaUrl.  

**Solution 1**
As solution, we can add a new method to interface to allow it to process schemas that has got from connectedInfo instead from a schemaUrl.

`public String processSchema( String schema, Util.PropertyList connectInfo) throws Exception;` 

Both of this methods have the same type of parameters, that means we cannot overload this method, we can add a extra parameter to allow differentiate or create a new method with a different name.

**Solution 2**
Another solution can be add a public method for pre-processing and another to post-processing.

Reply:
There's always solution 0. Use a stream as the API requires.